### ISSUE TITLE
intermittent publish error: Cannot read property 'main' of undefined 

### ISSUE THREAD

**Issue by dominictarr** [npm/npm#8772]
This error is intermittent, I try pub again and it works the second time!
Seems like a race condition where sometimes it is looking for `pkg.main`
but the package.json hasn't been read yet? (by the way, I never use a `.main` field, always just have `index.js`)

```
git push origin master && npm pub
...[gitstuff]...
npm ERR! Linux 4.0.4-2-ARCH
npm ERR! argv "/usr/bin/iojs" "/usr/bin/npm" "pub"
npm ERR! node v2.3.1
npm ERR! npm  v3.0.0

npm ERR! Cannot read property 'main' of undefined
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/dominic/c/sodiumperf/npm-debug.log
[dominic@localhost sodiumperf]$ npm pub
npm http request PUT https://registry.npmjs.org/sodiumperf
npm http 200 https://registry.npmjs.org/sodiumperf
- |----------------------------------
```

**Comment by calebmer** [comment]
I also get this when trying to install a dependency that points to a file in my package.json. For me running again does not work the second time.

**Comment by dominictarr** [comment]
@calebmer that is probably a very different problem, you should post another issue, and include your package.json & the dep that causes your problem.

**Comment by iarna** [comment]
@dominictarr Freaky– if you see this again can you snag the npm-debug.log?

**Comment by dominictarr** [comment]
here you go
```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/iojs', '/usr/bin/npm', 'pub' ]
2 info using npm@3.0.0
3 info using node@v2.3.1
4 verbose publish [ '.' ]
5 silly cache add args [ '.', null ]
6 verbose cache add spec .
7 silly cache add parsed spec { raw: '.',
7 silly cache add   scope: null,
7 silly cache add   name: null,
7 silly cache add   rawSpec: '.',
7 silly cache add   spec: '/home/dominic/c/sodiumperf',
7 silly cache add   type: 'directory' }
8 verbose addLocalDirectory /home/dominic/.npm/sodiumperf/0.0.1/package.tgz not in flight; packing
9 info prepublish sodiumperf@0.0.1
10 verbose tar pack [ '/home/dominic/.npm/sodiumperf/0.0.1/package.tgz',
10 verbose tar pack   '/home/dominic/c/sodiumperf' ]
11 verbose tarball /home/dominic/.npm/sodiumperf/0.0.1/package.tgz
12 verbose folder /home/dominic/c/sodiumperf
13 verbose stack TypeError: Cannot read property 'main' of undefined
13 verbose stack     at BundledPacker.Packer.applyIgnores (/usr/lib/node_modules/npm/node_modules/fstream-npm/fstream-npm.js:108:30)
13 verbose stack     at BundledPacker.applyIgnores (/usr/lib/node_modules/npm/lib/utils/tar.js:132:40)
13 verbose stack     at BundledPacker.<anonymous> (/usr/lib/node_modules/npm/node_modules/fstream-ignore/ignore.js:61:15)
13 verbose stack     at emitTwo (events.js:87:13)
13 verbose stack     at BundledPacker.emit (events.js:172:7)
13 verbose stack     at FileReader.<anonymous> (/usr/lib/node_modules/npm/node_modules/fstream/lib/dir-reader.js:131:12)
13 verbose stack     at emitOne (events.js:77:13)
13 verbose stack     at FileReader.emit (events.js:169:7)
13 verbose stack     at go (/usr/lib/node_modules/npm/node_modules/fstream/lib/reader.js:216:12)
13 verbose stack     at go (/usr/lib/node_modules/npm/node_modules/fstream/lib/reader.js:217:7)
14 verbose cwd /home/dominic/c/sodiumperf
15 error Linux 4.0.4-2-ARCH
16 error argv "/usr/bin/iojs" "/usr/bin/npm" "pub"
17 error node v2.3.1
18 error npm  v3.0.0
19 error Cannot read property 'main' of undefined
20 error If you need help, you may report this error at:
20 error     <https://github.com/npm/npm/issues>
21 verbose exit [ 1, true ]

```

**Comment by Goodmind** [comment]
I have this issue too

**Comment by thepian** [comment]
0 info it worked if it ends with ok
1 verbose cli [ 'node', '/usr/local/bin/npm', 'publish' ]
2 info using npm@2.12.0
3 info using node@v0.10.36
4 verbose publish [ '.' ]
5 silly cache add args [ '.', null ]
6 verbose cache add spec .
7 silly cache add parsed spec { raw: '.',
7 silly cache add   scope: null,
7 silly cache add   name: null,
7 silly cache add   rawSpec: '.',
7 silly cache add   spec: '/Users/Shared/Projects/Websites/libs/ss-sockjs',
7 silly cache add   type: 'directory' }
8 verbose addLocalDirectory /Users/hvendelbo/.npm/ss-sockjs/0.1.3/package.tgz not in flight; packing
9 verbose tar pack [ '/Users/hvendelbo/.npm/ss-sockjs/0.1.3/package.tgz',
9 verbose tar pack   '/Users/Shared/Projects/Websites/libs/ss-sockjs' ]
10 verbose tarball /Users/hvendelbo/.npm/ss-sockjs/0.1.3/package.tgz
11 verbose folder /Users/Shared/Projects/Websites/libs/ss-sockjs
12 info prepublish ss-sockjs@0.1.3
13 verbose type non_object_property_load
14 verbose stack TypeError: Cannot read property 'main' of undefined
14 verbose stack     at Packer.applyIgnores (/usr/local/lib/node_modules/npm/node_modules/fstream-npm/fstream-npm.js:108:30)
14 verbose stack     at Packer.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/fstream-npm/node_modules/fstream-ignore/ignore.js:61:15)
14 verbose stack     at Packer.emit (events.js:98:17)
14 verbose stack     at FileReader.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/fstream/lib/dir-reader.js:131:12)
14 verbose stack     at FileReader.emit (events.js:95:17)
14 verbose stack     at go (/usr/local/lib/node_modules/npm/node_modules/fstream/lib/reader.js:216:12)
14 verbose stack     at go (/usr/local/lib/node_modules/npm/node_modules/fstream/lib/reader.js:217:7)
14 verbose stack     at statCb (/usr/local/lib/node_modules/npm/node_modules/fstream/lib/reader.js:218:7)
14 verbose stack     at Object.oncomplete (evalmachine.<anonymous>:108:15)
15 verbose cwd /Users/Shared/Projects/Websites/libs/ss-sockjs
16 error Darwin 14.4.0
17 error argv "node" "/usr/local/bin/npm" "publish"
18 error node v0.10.36
19 error npm  v2.12.0
20 error Cannot read property 'main' of undefined
21 error If you need help, you may report this error at:
21 error     <https://github.com/npm/npm/issues>
22 verbose exit [ 1, true ]

**Comment by parro-it** [comment]
I had this problem too. It appear that I put my module main js file in .npmignore.
I removed it and it worked fine.

**Comment by iarna** [comment]
Oh hey, we fixed this with npm@3.1.1 and npm@2.13.1 w/ pr #8858  https://github.com/npm/fstream-npm/commit/e831c7c4c82acffc80e7c7e86ae2c456e8bc0c0c !

**Comment by dominictarr** [comment]
thanks!
